src/session/key.ts

```diff
@@ -0,0 +1,126 @@
+  DELETE_DATA_SET_TYPEHASH,
+]
+
+export const PDP_PERMISSION_NAMES: Record<string, string> = {
+  [CREATE_DATA_SET_TYPEHASH]: 'CreateDataSet',
+  [ADD_PIECES_TYPEHASH]: 'AddPieces',
+  [SCHEDULE_PIECE_REMOVALS_TYPEHASH]: 'SchedulePieceRemovals',
+  [DELETE_DATA_SET_TYPEHASH]: 'DeleteDataSet',
+}
+
+export class SessionKey {
+  private readonly _provider: ethers.Provider
+  private readonly _registry: ethers.Contract
+  private readonly _signer: ethers.Signer
+  private readonly _owner: ethers.Signer
+  public readonly expiries: Record<string, bigint | null>
+
+  public constructor(
+    provider: ethers.Provider,
+    sessionKeyRegistry: ethers.Contract,
+    signer: ethers.Signer,
+    owner: ethers.Signer
+  ) {
+    this._provider = provider
+    this._registry = sessionKeyRegistry.connect(owner) as ethers.Contract
+    this._signer = signer
+    this._owner = owner
+    this.expiries = {
+      CREATE_DATA_SET_TYPEHASH: null,
+      ADD_PIECES_TYPEHASH: null,
+      SCHEDULE_PIECE_REMOVALS_TYPEHASH: null,
+      DELETE_DATA_SET_TYPEHASH: null,
+    }
+  }
+
+  getSigner(): ethers.Signer {
+    return this._signer
+  }
+
+  /**
+   * Refreshes SessionKey.expiries by fetching the expiries of those permissions from the registry
+   */
+  async fetchExpiries(): Promise<void> {
+    const network = await getFilecoinNetworkType(this._provider)
+
+    const multicall = new ethers.Contract(
+      CONTRACT_ADDRESSES.MULTICALL3[network],
+      CONTRACT_ABIS.MULTICALL3,
+      this._provider
+    )
+    const registryInterface = new ethers.Interface(CONTRACT_ABIS.SESSION_KEY_REGISTRY)
+
+    const [ownerAddress, signerAddress, registryAddress] = await Promise.all([
+      this._owner.getAddress(),
+      this._signer.getAddress(),
+      this._registry.getAddress(),
+    ])
+
+    // Prepare multicall batch
+    const calls: Array<{ target: string; allowFailure: boolean; callData: string }> = []
+    for (const permission of PDP_PERMISSIONS) {
+      calls.push({
+        target: registryAddress,
+        allowFailure: true,
+        callData: registryInterface.encodeFunctionData('authorizationExpiry', [
+          ownerAddress,
+          signerAddress,
+          permission,
+        ]),
+      })
+    }
+
+    // Execute multicall
+    const results = await multicall.aggregate3.staticCall(calls)
+
+    for (let i = 0; i < PDP_PERMISSIONS.length; i++) {
+      this.expiries[PDP_PERMISSIONS[i]] = registryInterface.decodeFunctionResult(
+        'authorizationExpiry',
+        results[i].returnData
+      )[0]
+    }
+  }
+
+  /**
+   * Owner authorizes signer with permissions until expiry
+   * @param expiry unix time (block.timestamp) that the permissions expire
+   * @param permissions list of permissions granted to the signer
+   */
+  async login(expiry: bigint, permissions: string[] = PDP_PERMISSIONS): Promise<ethers.TransactionResponse> {
+    return await this._registry.login(await this._signer.getAddress(), expiry, permissions)
+  }
```

Comment:
this contract call registers public key but also refreshes expiry/permissions if already registered ?

Reply:
yes, it can be used to create, renew, or invalidate.